backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PoolPostgresExecutor.java

```diff
@@ -0,0 +1,107 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.postgres.utils;
+
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.MAX_RETRY_ATTEMPTS;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.MIN_BACKOFF;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.PGSQL_DIALECT;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.SETTINGS;
+
+import java.time.Duration;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.inject.Inject;
+
+import org.apache.james.lifecycle.api.Disposable;
+import org.jooq.DSLContext;
+import org.jooq.Record;
+import org.jooq.impl.DSL;
+import org.reactivestreams.Publisher;
+
+import io.r2dbc.pool.ConnectionPool;
+import io.r2dbc.pool.ConnectionPoolConfiguration;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.R2dbcBadGrammarException;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
+
+public class PoolPostgresExecutor implements Disposable {
+
+    static int INITIAL_SIZE = 10;
+    static int MAX_SIZE = 50;
+    static Duration MAX_IDLE_TIME = Duration.ofMillis(5000);
+    private final ConnectionPool pool;
+
+    @Inject
+    public PoolPostgresExecutor(ConnectionFactory connectionFactory) {
+        ConnectionPoolConfiguration configuration = ConnectionPoolConfiguration.builder(connectionFactory)
+            .maxIdleTime(MAX_IDLE_TIME)
+            .initialSize(INITIAL_SIZE)
+            .maxSize(MAX_SIZE)
+            .build();
+        pool = new ConnectionPool(configuration);
+    }
+
+    public Mono<Void> executeVoid(Function<DSLContext, Mono<?>> queryFunction) {
+        return Mono.usingWhen(pool.create(),
+            connection -> dslContext(connection)
+                .flatMap(queryFunction)
+                .retryWhen(Retry.backoff(MAX_RETRY_ATTEMPTS, MIN_BACKOFF)
+                    .filter(preparedStatementConflictException()))
```

Comment:
this exception could not occur when we use connection pool

Reply:
Can you explain more?
Because `Mono.usingWhen`?

Reply:
the preparedStatementConflictException only occur when we use a single connection in multiple threads. In case of connection pool, it is not necessary to handle this case

Reply:
hmm... 
Before, we shared a same connection between threads, which led to prepare statements conflict: thread A and thread B tried to prepare different statements with a same statement ID -> preparedStatementConflictException.

Maybe Hung idea is that thanks to connection pool, a connection is no longer shared between thread at a same time -> the conflict could not happen?

Reply:
Well this implementation of a pool is sub-optimal as our workload allows some amount of connection sharing.

Maximum pool size put a high boundary on concurrency that we were not needing before.

TBH I would be happier with a 1 connection per thread model, where we simply apply some `least request` decision mechanism (ie count requests on each open connections and choose the one with the least...

Reply:
>  the conflict could not happen?

We can remove it temporarily, and monitor it?